mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp

```diff
@@ -2165,109 +2196,196 @@ struct GridwiseAttentionAccelRewritePattern
+          Value gemm1MNThreadwiseView =
+              transform(rewriter, gemm1OutBufferPerG1MBlock,
+                        invertedGemm1threadSubTileMaps);
+          // Rescale/correct output, rowMax and rowSums
+          Value attentionOutAccBufferView =
+              transform(rewriter, attentionOutAccBufferPerG1MBlock,
+                        attentionOutAccBufferThreadSubTileViewMaps);
+          createAttentionRowStateCorrections(
+              rewriter, loc, gemm1MNThreadwiseView, attentionOutAccBufferView,
+              expMaxDiffRowBuffer);
+        }
+      }
+    }
+    {
+      affine::AffineForOp g1MLoopOp =
+          rewriter.create<affine::AffineForOp>(loc, 0, gemm1MBlocks, 1);
+      {
+        OpBuilder::InsertionGuard guard(rewriter);
+        rewriter.setInsertionPointToStart(g1MLoopOp.getBody());
+        Value g1MLoopIndVar = g1MLoopOp.getInductionVar();
+        Value attentionOutAccBufferPerG1MBlock = attentionOutAccBuffer;
+        if (gemm1MBlocks > 1) {
+          attentionOutAccBufferPerG1MBlock = createSliceOfFirstDim(
+              rewriter, loc, attentionOutAccBuffer, g1MLoopIndVar);
+        }
+        Value attentionOutAccBufferView =
+            transform(rewriter, attentionOutAccBufferPerG1MBlock,
+                      attentionOutAccBufferThreadSubTileViewMaps);
+        scaleFinalOutput(rewriter, loc, attentionOutAccBufferView,
+                         sumRowBuffer);
       }
     }
-    scaleFinalOutput(rewriter, loc, attentionOutAccBufferView, sumRowBuffer);
     if (elemTypeQxK != elemTypeOut) {
       createTypeConversionLaGeneric(rewriter, loc, attentionOutAccBuffer,
                                     attentionOutAccBufferOutTyped);
     }
 #ifdef ROCK_DEBUG_ATTENTION_REMOVE_SOFTMAX
     attentionOutAccBufferOutTyped = gemm1OutBuffer;
 #endif
+    // We flatten output buffer in case gemm1MBlocks > 1
+    // where those are iterated.
+    Value attentionOutAccBufferOutTypedFlat = attentionOutAccBufferOutTyped;
+    MemRefType attentionOutAccBufferOutType =
+        attentionOutAccBufferOutTyped.getType().cast<MemRefType>();
+    int64_t numElementsAttnOut = attentionOutAccBufferOutType.getNumElements();
+    if (attentionOutAccBufferOutType.getRank() > 1) {
+      Type attentionOutAccBufferOutTypedElType =
+          attentionOutAccBufferOutType.getElementType();
+      auto attentionOutAccBufferOutTypedFlatType = MemRefType::get(
+          {numElementsAttnOut}, attentionOutAccBufferOutTypedElType,
+          AffineMap{}, privateMemoryAddressSpace);
+      auto reassociation =
+          getReassociationForFlattening(attentionOutAccBufferOutType);
+      attentionOutAccBufferOutTypedFlat =
+          rewriter.create<memref::CollapseShapeOp>(
+              loc, attentionOutAccBufferOutTypedFlatType,
+              attentionOutAccBufferOutTyped, reassociation);
+    }
+    // This map with create upper view [gblock, nblock, flatiter] -> [gblock,
```

Comment:
Is this a typo? "This map will create an upper view ..."

Reply:
thanks for catching this.